src/Pact/Types/Persistence.hs

```diff
@@ -120,32 +156,90 @@ makeLenses ''ModuleData
-instance (ToJSON r, FromJSON r) => FromJSON (ModuleData r) where
+-- On chain this encoded as follows:
+-- * refMap: object
+-- * dependencies: Array
+
+instance ToJSON r => ToJSON (ModuleData r) where
+  toJSON o = enableToJSON "Pact.Types.Persistence.ModuleData r" $ A.Object $ mconcat
+    [ "dependencies" .?= if HM.null (_mdDependencies o)
+      then Nothing
+      else Just (LHM.toList $ legacyHashMap_ $ _mdDependencies o)
+    , "module" .= _mdModule o
+    , "refMap" .= _mdRefMap o
+    ]
+  toEncoding o = pairs $ mconcat
+    [ "dependencies" .?= if HM.null (_mdDependencies o)
+      then Nothing
+      else Just (LHM.toList $ legacyHashMap_ $ _mdDependencies o)
+    , "module" .= _mdModule o
+    , "refMap" .= legacyHashMap (_mdRefMap o)
+    ]
+  {-# INLINE toJSON #-}
+  {-# INLINE toEncoding #-}
+
+instance (J.Encode r, Eq r) => J.Encode (ModuleData r) where
+  build o = J.object
+    [ "dependencies" J..??= J.Array (J.Array <$> LHM.toList (legacyHashMap_ (_mdDependencies o)))
+    , "module" J..= _mdModule o
+    , "refMap" J..= legacyHashMap (_mdRefMap o)
+    ]
+  {-# INLINE build #-}
+
+instance (FromJSON r) => FromJSON (ModuleData r) where
   parseJSON =
     withObject "ModuleData" $ \o ->
       ModuleData
       <$> o .: "module"
       <*> o .: "refMap"
-      <*> (fromMaybe HM.empty <$> o .:? "dependencies")
+      <*> (HM.fromList <$> (fromMaybe mempty <$> o .:? "dependencies"))
+
+instance Arbitrary r => Arbitrary (ModuleData r) where
+  arbitrary = ModuleData
+    <$> arbitrary
+    <*> scale (min 10) arbitrary
+    <*> scale (min 10) arbitrary
+
+-- -------------------------------------------------------------------------- --
+-- PersistModuleData
 
 type PersistModuleData = ModuleData (Ref' PersistDirect)
 
 instance ToJSON (Ref' PersistDirect) where
-  toJSON (Ref t) = object [ "ref" .= t ]
-  toJSON (Direct pd) = object [ "direct" .= pd ]
+  toJSON = enableToJSON "Pact.Types.Persistence.ModuleData r" . object . \case
+    (Ref t) -> [ "ref" .= t ]
+    (Direct pd) -> [ "direct" .= pd ]
+
+  toEncoding (Ref t) = pairs ("ref" .= t)
+  toEncoding (Direct pd) = pairs ("direct" .= pd)
+
+  {-# INLINE toJSON #-}
+  {-# INLINE toEncoding #-}
+
+instance J.Encode (Ref' PersistDirect) where
+  build (Ref t) = J.object ["ref" J..= t]
+  build (Direct pd) = J.object ["direct" J..= pd]
+  {-# INLINE build #-}
 
 instance FromJSON (Ref' PersistDirect) where
   parseJSON v =
     withObject "Ref" (\o -> Ref <$> o .: "ref") v <|>
     withObject "Direct" (\o -> Direct <$> o .: "direct") v
+  {-# INLINE parseJSON #-}
 
+-- -------------------------------------------------------------------------- --
+-- RowKey
 
 -- | Row key type for user tables.
 newtype RowKey = RowKey Text
-    deriving (Eq,Ord,IsString,ToTerm,AsString,Show,Pretty,Generic,NFData)
+    deriving (Eq,Ord,Generic)
+    deriving newtype (IsString,ToTerm,AsString,Show,Pretty,NFData)
```

Comment:
This doesn't change the way instances are derived, because `newtype` deriving was used for all those instances by default. 

This change just makes it explicit what mechanism is used and makes sure that it can't change accidentally when other deriving mechanisms are enabled in the module by adding the respective pragmas.